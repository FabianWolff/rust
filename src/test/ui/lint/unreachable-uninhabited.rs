// Regression test for #89779.

#![forbid(unreachable_code)]
//~^ NOTE: the lint level is defined here

fn stop() -> Result<std::convert::Infallible, u64> {
    Err(5)
}

fn _foo1() {
    if false {
        stop().unwrap();
    }

    println!("Hello, world!");
    // no warnings here
}

fn _foo2() {
    if false {
    //~^ NOTE: any code following this expression is unreachable
    //~| NOTE: both branches of this `if` expression diverge
        stop().unwrap();
    } else {
        stop().unwrap();
    }

    println!("Hello, world!");
    //~^ ERROR: unreachable expression
}

fn _foo3() {
    match stop() {
    //~^ NOTE: any code following this expression is unreachable
    //~| NOTE: all arms of this `match` expression diverge
        Ok(x) => stop().unwrap(),
        Err(_) => stop().unwrap(),
    };

    println!("Hello, world!");
    //~^ ERROR: unreachable expression
}

fn bar() -> Result<std::convert::Infallible, u64> {
    Err(5)
}

fn main() {
    match bar().expect("should") {}
    // no warnings here
}
