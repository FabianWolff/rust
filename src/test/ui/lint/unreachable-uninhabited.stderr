error: unreachable expression
  --> $DIR/unreachable-uninhabited.rs:28:5
   |
LL | /     if false {
LL | |
LL | |
LL | |         stop().unwrap();
LL | |     } else {
LL | |         stop().unwrap();
LL | |     }
   | |_____- any code following this expression is unreachable
LL | 
LL |       println!("Hello, world!");
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable expression
   |
note: the lint level is defined here
  --> $DIR/unreachable-uninhabited.rs:3:11
   |
LL | #![forbid(unreachable_code)]
   |           ^^^^^^^^^^^^^^^^
note: both branches of this `if` expression diverge
  --> $DIR/unreachable-uninhabited.rs:20:5
   |
LL | /     if false {
LL | |
LL | |
LL | |         stop().unwrap();
LL | |     } else {
LL | |         stop().unwrap();
LL | |     }
   | |_____^
   = note: this error originates in the macro `$crate::format_args_nl` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unreachable expression
  --> $DIR/unreachable-uninhabited.rs:40:5
   |
LL | /     match stop() {
LL | |
LL | |
LL | |         Ok(x) => stop().unwrap(),
LL | |         Err(_) => stop().unwrap(),
LL | |     };
   | |_____- any code following this expression is unreachable
LL | 
LL |       println!("Hello, world!");
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable expression
   |
note: all arms of this `match` expression diverge
  --> $DIR/unreachable-uninhabited.rs:33:5
   |
LL | /     match stop() {
LL | |
LL | |
LL | |         Ok(x) => stop().unwrap(),
LL | |         Err(_) => stop().unwrap(),
LL | |     };
   | |_____^
   = note: this error originates in the macro `$crate::format_args_nl` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

